### PR TITLE
These definitions have been deprecated by the vsts maintenance jobs.

### PR DESCRIPTION
**Pull request by JohnTortugo**
Related to issue: https://github.com/dotnet/core-eng/issues/2964

### CODE REVIEW

**Comment by JohnTortugo** [comment]
@weshaggard, assuming all checks pass, is it OK to merge this PR into release branch?

**Comment by weshaggard** [comment]
CI checks won't verify these changes and we are in lock-down with stable builds so I'd like to hold on merging this for now. 

@wtgodbe once we re-open these branches can you merge this?

File: buildpipeline/DotNet-CoreFx-Trusted-OSX.json [hunk 1]
```diff
@@ -1,5 +1,29 @@
 {
   "build": [
+    {
+      "environment": {},
+      "enabled": true,
+      "continueOnError": true,
+      "alwaysRun": false,
+      "displayName": "Run AgentTools/Begin.ps1",
```
Comment (MattGal) [inline buildpipeline/DotNet-CoreFx-Trusted-OSX.json]: OSX should have begin/end.sh files

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-OSX.json]: Fixed. Updated OSX and Linux definitions.

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json [hunk 2]
```diff
@@ -1,5 +1,29 @@
...
+    {
+      "environment": {},
+      "enabled": true,
+      "continueOnError": true,
+      "alwaysRun": false,
+      "displayName": "Run AgentTools/Begin.sh",
+      "timeoutInMinutes": 0,
+      "condition": "succeededOrFailed()",
+      "task": {
+        "id": "e213ff0f-5d5c-4791-802d-52ea3e7be1f1",
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: I believe this is the PS task id, you will need to change it, and the script code will need to be changed.

**Comment by weshaggard** [comment]
I chatted with @wtgodbe offline and he is going to submit a scoped verison of this PR to unblock the current builds .

**Comment by wtgodbe** [comment]
We'll actually have to wait on merging this until branches are open again. I've put up https://github.com/dotnet/corefx/pull/28735 to unblock official builds - @JohnTortugo please rebase against that PR once I merge it

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json (hunk 2 above)
Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: I am not sure what you meant by changing the script code. You mean to update the "script": entry below to use bash commands, right?

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: As per @jcagme suggestion I added a new file, called extensionTask.sh, that can be used in place of begin.sh / end.sh. This bash script receive as parameter the name of the extension task: "begin" or "end".

**Comment by JohnTortugo** [comment]
As per @jcagme suggestion I added a new file, called extensionTask.sh, that can be used in place of begin.sh / end.sh. This bash script receive as parameter the name of the extension task: "begin" or "end".

Any feedback on that @weshaggard ?

**Comment by weshaggard** [comment]
>As per @jcagme suggestion I added a new file, called extensionTask.sh, that can be used in place of begin.sh / end.sh. This bash script receive as parameter the name of the extension task: "begin" or "end".

What does that buy us? We still need to check for the existence of that script before you execute it.

File: src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509chain.cpp [hunk 3]
```diff
@@ -154,7 +154,8 @@ static void MergeStatusCodes(CFTypeRef key, CFTypeRef value, void* context)
     CFStringRef keyString = reinterpret_cast<CFStringRef>(key);
 
     if (CFEqual(keyString, CFSTR("NotValidBefore")) || CFEqual(keyString, CFSTR("ValidLeaf")) ||
-        CFEqual(keyString, CFSTR("ValidIntermediates")) || CFEqual(keyString, CFSTR("ValidRoot")))
```
Comment (weshaggard) [inline src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509chain.cpp]: This isn't part of your change .

File: extensionTask.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+
+if [ "$1" == "begin" ]; then
```
Comment (weshaggard) [inline extensionTask.sh]: This script will not be accessible in the build definitions until the cloning has happened so we cannot depend on that. Also I don't like checking this into every repo (especially in the root of the repo).

Comment (weshaggard) [reply extensionTask.sh]: I suggest using an inline bash script for the begin/end steps.

Comment (jcagme) [reply extensionTask.sh]: I just did not find a CMD task where one could write the script inline, hence my recommendation. If there is one where this can be done, we should use it by all means

Comment (jcagme) [reply extensionTask.sh]: Oh this is cool!Thanks for the pointer

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json [hunk 5]
```diff
@@ -1,5 +1,28 @@
...
+        "versionSpec": "0.*",
+        "definitionType": "task"
+      },
+      "inputs": {
+        "type": "InlineScript",
+        "scriptPath": "",
+        "args": "",
+        "cwd": "",
+        "failOnStandardError": "false",
+        "script": "if [ -f \"Begin.sh\" ]; then ./Begin.sh ; else echo \"Begin.sh script does not exist. Moving on.\" ; fi"
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: What about AgentToolsPath?

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json [hunk 6]
```diff
@@ -1,5 +1,28 @@
...
+      "enabled": true,
+      "continueOnError": true,
+      "alwaysRun": false,
+      "displayName": "Run Begin.sh",
+      "timeoutInMinutes": 0,
+      "condition": "succeeded()",
+      "refName": "TaskRunBegin",
+      "task": {
+        "id": "10f1f9a1-74b0-47ab-87bf-e3c9c68e8b0d",
+        "versionSpec": "0.*",
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: Is that the correct version spec? How are you testing these changes? Can you please create a sample definition and verify that these tasks work?

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: @jcagme told me that all I had to do was to paste this on the file :-/ ; Thanks for the suggestion. I am going to create a sample definition.

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: @jcagme helped me to test it: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build?_a=summary&buildId=1552684

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json (hunk 5 above)
Comment (weshaggard) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: I still think you need to take AgentToolsPath into account like with the ps1 otherwise we will not be able to control where these scripts are on the agent machines.

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json (hunk 6 above)
Comment (weshaggard) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: Thanks. Now please take into account the AgentToolsPath and you should be set. I suggest testing it on a definition where you can put a script there to verify it gets discovered and ran when it is present as well.

Comment (jcagme) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: Yes, we need to add $(AgentToolsPath) to the script so when we deploy the files to the machines this will actually execute what we attempt to execute

Comment (jcagme) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: You can use the branch/build definition used yesterday to drop a test script. Probably you already did this since you asked me this yesterday :)

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: What's the name of this var in Linux/OSX? I tried $AgentToolsPath and $AGENTTOOLSPATH but no success.

Comment (jcagme) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: What do you mean by no success? How does it fail?

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: It seems that these variables don't exist or are empty.

Comment (jcagme) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: @ilyas1974 is there a path already defined for AgentToolsPath in Mac/Linux? As @JohnTortugo points out there is "AGENT_TOOLSDIRECTORY=/Users/buildagent/agent/_work/_tool". Do you recommend using this or we should create a new one?

Comment (weshaggard) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: @jcagme I thought the plan was to pass in the value in the definition because we didn't have a great way to set them on the agent machine. Is that not the case?

Comment (jcagme) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: We agreed on adding this environment variable on agent imaging. This is currently what happens on Windows agents and doing the same for Linux/Mac should be straight forward. @ilyas1974 ?

Comment (weshaggard) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: @jcagme it could be a system wide environment variable or a property passed in through the root build definition. Either way it should be consistently named AgentToolsDirectory.

Comment (jcagme) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: Yes, it will be. When @ilyas1974 did the work on Windows agents everything was documented so the same work done on different agents (which ever the OS) will be the same.

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: Hey, @JC, is there anything I can do to get this moving again?

Comment (jcagme) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: Sent @ilyas1974 an email since I don't know if he gets notifications about tags here

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: @jcagme talked to @ilyas1974 about adding this new environment variable in Linux / OSX. The request is being tracked here: https://devdiv.visualstudio.com/web/wi.aspx?pcguid=011b8bdf-6d56-4f87-be0d-0092136884d9&id=597942

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: @ilyas1974 said that due his team current demands it may take until the end of the month to get the env. vars on Linux / OSX set. I spoke with @jcagme and we are considering moving these PR forward without the begin/end.sh scripts on Linux / OSX. 

I am going to start testing the changes on Windows. @weshaggard , does that sounds good to you?

Comment (weshaggard) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: It is fine if the env variables aren't on the machines yet. You can test these by setting the property when you queue the build.

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: Hi @weshaggard , I've tested this on Windows, Linux and OSX now. I uploaded an executable script from a Linux VM and tested, it worked. I'd advise to still keep the "chmod +x" command before trying to execute the script, though. Just in case someone forget to +x it before uploading, or upload it from a Windows box.

Build where I tested it: 
https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/index?buildId=1586143&_a=summary 
[Linux] https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build?_a=summary&buildId=1586178
[Windows] https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build?_a=summary&buildId=1586161
[Windows] https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build?_a=summary&buildId=1586158
[OSX] https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build?_a=summary&buildId=1586157
[Linux] https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build?_a=summary&buildId=1586154

If this sounds good for you I'll apply the patch on the other PRs.

Comment (weshaggard) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: Thanks for testing but unless I'm missing something I don't see those updates to this PR to review. 

As far as I can tell from your test runs you didn't even call chmod:
```
if [ -f "$AGENTTOOLSPATH/begin.sh" ]; then ls -la $AGENTTOOLSPATH/begin.sh ; $AGENTTOOLSPATH/begin.sh ; else echo "$AGENTTOOLSPATH/begin.sh script does not exist. Moving on." ; fi
```

At any rate update the PR with what you have tested and I will take a closer review.

Comment (JohnTortugo) [reply buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: That was tested in another branch. I didn't include the chmod +x because, as you asked, I uploaded the file (begin.sh) from a Linux machine after locally setting chmod on it. I'll replicate the buildefinition changes on this PR.

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json [hunk 7]
```diff
@@ -20,7 +20,7 @@
         "args": "",
         "cwd": "",
         "failOnStandardError": "false",
-        "script": "if [ -f \"Begin.sh\" ]; then ./Begin.sh ; else echo \"Begin.sh script does not exist. Moving on.\" ; fi"
+        "script": "if [ -f \"$AGENTTOOLSPATH/begin.sh\" ]; then $AGENTTOOLSPATH/begin.sh ; else echo \"$AGENTTOOLSPATH/begin.sh script does not exist. Moving on.\" ; fi"
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: note that linux is case sensitive so we we should be sure we want lowercase.

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json [hunk 8]
```diff
@@ -505,6 +505,9 @@
     },
     "PB_CleanAgent": {
       "value": "true"
+    },
+    "AgentToolsPath": {
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: Do we really want to default this to `$(Build.SourcesDirectory)`? I think I would either default it to empty.

File: buildpipeline/DotNet-CoreFx-Trusted-Windows-NoTest.json [hunk 9]
```diff
@@ -1,5 +1,29 @@
...
+      "task": {
+        "id": "e213ff0f-5d5c-4791-802d-52ea3e7be1f1",
+        "versionSpec": "2.*",
+        "definitionType": "task"
+      },
+      "inputs": {
+        "targetType": "inline",
+        "filePath": "",
+        "arguments": "",
+        "script": "if (Test-Path \"$(AgentToolsPath)\\Begin.ps1\") {\n   \"Begin.ps1 script found. Executing...\"\n   cd $(AgentToolsPath)\n   & $(AgentToolsPath)\\Begin.ps1\n} else {\n   \"Begin.ps1 script does not exist. Moving on...\"\n}",
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-Windows-NoTest.json]: nit: For consistency if we go with lowercase in the shell scripts I would suggest doing the same with the ps1 scripts. My preference is lowercase.

File: buildpipeline/DotNet-CoreFx-Trusted-Windows-NoTest.json [hunk 10]
```diff
@@ -1,5 +1,29 @@
 {
   "build": [
+    {
+      "environment": {},
+      "enabled": true,
+      "continueOnError": true,
+      "alwaysRun": false,
+      "displayName": "Run AgentTools/Begin.ps1",
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-Windows-NoTest.json]: for consistency "run begin.ps1"

File: buildpipeline/DotNet-CoreFx-Trusted-OSX.json [hunk 11]
```diff
@@ -241,6 +264,29 @@
...
       }
+    },
+    {
+      "environment": {},
+      "enabled": true,
+      "continueOnError": true,
+      "alwaysRun": false,
+      "displayName": "Run end.sh",
+      "timeoutInMinutes": 0,
+      "condition": "succeeded()",
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-OSX.json]: the ps scripts are set to run `succeededOrFailed` which I think is a good idea for these as well.

File: buildpipeline/DotNet-Trusted-Publish-Symbols.json [hunk 12]
```diff
@@ -1,5 +1,29 @@
 {
   "build": [
+    {
+      "environment": {},
+      "enabled": true,
+      "continueOnError": true,
+      "alwaysRun": false,
+      "displayName": "Run AgentTools/Begin.ps1",
+      "timeoutInMinutes": 0,
+      "condition": "succeededOrFailed()",
```
Comment (weshaggard) [inline buildpipeline/DotNet-Trusted-Publish-Symbols.json]: After reading https://docs.microsoft.com/en-us/vsts/build-release/concepts/process/conditions?view=vsts I suggest we use condition "always" for both begin and end scripts.

**Comment by weshaggard** [comment]
I added a few more clean-up style comments but otherwise seems reasonable. 

I would suggest we get these changes into an active master branch first and then do some testing to ensure they work as expected (with and without a script on the agent machine) before committing them to all our build definitions.

**Comment by JohnTortugo** [comment]
@weshaggard , FYI I pushed new commit with the changes you suggested. Thanks!

File: buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json [hunk 13]
```diff
@@ -1,5 +1,28 @@
 {
   "build": [
+    {
+      "environment": {},
+      "enabled": true,
+      "continueOnError": true,
+      "alwaysRun": false,
```
Comment (weshaggard) [inline buildpipeline/DotNet-CoreFx-Trusted-Linux-Crossbuild.json]: We should understand what the difference between "alwaysRun"= false and condition always() and be sure the don't conflict with each other.

File: buildpipeline/DotNet-CoreFx-Trusted-Windows-NoTest.json (hunk 9 above)
Comment (weshaggard) [reply buildpipeline/DotNet-CoreFx-Trusted-Windows-NoTest.json]: I see you updated the name of the task but not the name of the scripts to lowercase. I would suggest doing both for consistency.

Comment (weshaggard) [reply buildpipeline/DotNet-CoreFx-Trusted-Windows-NoTest.json]: Looks like you did do it for end.ps1 just missed the Begin.ps1 scripts in both definitions.

File: src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509chain.cpp [hunk 14]
```diff
@@ -154,8 +154,7 @@ static void MergeStatusCodes(CFTypeRef key, CFTypeRef value, void* context)
     CFStringRef keyString = reinterpret_cast<CFStringRef>(key);
 
     if (CFEqual(keyString, CFSTR("NotValidBefore")) || CFEqual(keyString, CFSTR("ValidLeaf")) ||
-        CFEqual(keyString, CFSTR("ValidIntermediates")) || CFEqual(keyString, CFSTR("ValidRoot")) ||
```
Comment (weshaggard) [inline src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509chain.cpp]: This change is unrelated. I suggest rebasing on top of the branch.

**Comment by JohnTortugo** [comment]
The rebase for this was getting me crazy - too many conflicting commits. I found it easier to create another PR: https://github.com/dotnet/corefx/pull/29104
